TLS 1.3 draft-23 support

kaduk opened the pull request:
The basic renumbering stuff is there, but I think there may be some more simplification possible for the RSA-PSS key OID split, and of course several more tests are likely in order.

kaduk commented on test/ssl-tests/20-cert-select.conf.in:
```diff
@@ -87,7 +87,7 @@ our @tests = (
         },
         test   => {
-            "ExpectedServerCertType" =>, "RSA-PSS",
+            "ExpectedServerCertType" =>, "RSA",
```
Hmm, but something is not right here, since this is supposed to be using the ```$server_pss``` block, which actually should be an RSA-PSS key in the cert.  Need to dig further...

kaduk replied:
The story here is actually interesting, but in a depressing sort of way ... even when the client configuration directs it to use the rsa_pss_pss_sha256 sigalg, it actually ends up wending 0x0804 (rsa_pss_rsae_sha256) on the wire.  This is because tls1_set_sigalgs() gets an array of sigNID+hashNID pairs, and goes through the lookup table to find matching SIGALG_LOOKUP entries.  Now that the TLS spec has split the RSA ones up by key OID, there are two matching table elements, and we pick the first (wrong) one.  This sigNID+hashNID format is intrinsic to our public APIs SSL{,_CTX}_set1_sigalgs{,_list}(), so the correct behavior will need to be implemented as a new public API (and SSL_ctrl case?  See also #1319) and a decision made on which way the existing API should disambiguate.
Oh, and s_client/s_server need to learn to use the new API.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2291,6 +2308,24 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
+                lu = tls1_lookup_sigalg(s->s3->tmp.peer_cert_sigalgs[i]);
+                if (lu == NULL || !ssl_has_cert(s, lu->sig_idx))
+                    continue;
+                s->s3->tmp.cert = &s->cert->pkeys[sig_idx];
```
Do you mean `lu->sig_idx` here?

kaduk replied:
Yes...and that probably explains the issue with the tests.

kaduk replied:
(actually it doesn't impact the tests, since we don't generate the extension yet)

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2291,6 +2308,24 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
+            if (s->s3->tmp.cert == NULL)
+                SSLfatal(s, SSL_AD_HANDSHAKE_FAILURE, SSL_F_TLS_CHOOSE_SIGALG,
+                         SSL_R_NO_SUITABLE_SIGNATURE_ALGORITHM);
+                return 0;
```
Err...shouldn't this all be in `{}`? Even if you do that though I'm not sure why we then fall through to the "Look for a certificate matching shared sigalgs" code below. That seems to do a lot of work setting up `sig_idx` which we then ignore because `s->s3->tmp.cert != NULL`. Or did I misunderstand something?

kaduk replied:
We need to (re)compute ```lu``` using the signature_algorithms (non-_cert) extension, in order to set ```s->s3->tmp.sigalg``` to the proper value for signing TLS messages.  signature_algorithms_cert is *only* used for choosing what certificate (chain, but this currently doesn't look at the chain) to present.

tmshort replied:
+1 to mattcaswell; missing braces.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2363,7 +2398,8 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
 
         if (SSL_USE_SIGALGS(s)) {
-            if (s->s3->tmp.peer_sigalgs != NULL) {
+            if (s->s3->tmp.peer_cert_sigalgs != NULL
```
I'm slightly confused by this change. How does the value of `peer_cert_sigalgs` ever get used in the TLSv1.2 case?

kaduk replied:
In the present form of this pull request, the value does not get used in this block.  (I may have lost track that this is in the else clause of a TLS13 block, when writing it.)  The motivation was along the lines of that if the peer is making a claim about what signature algorithms to use, then we should not fall back to a default list if there is no overlap in the provided values.  That said, looking back at the change with fresh eyes, the practical effect is to forbid the case of sending signature_algorithms_cert and not signature_algorithms.  The spec does say "TLS 1.2 implementations SHOULD also process this extension [signature_algorithms_cert].", so it does seem spec-compliant for a client to only offer TLS 1.2, send signature_algorithms_cert, but request the default sigalgs by not sending the regular signature_algorithms.  That implies that this check should be removed, but the block to pick a cert should be copied from the TLS13 case and duplicated here.  (Given the conditions for when the default sigalgs should be used, I'm not sure that it's a win to try to consolidate the signature_algorithms_cert block.)

tmshort commented on include/openssl/tls1.h:
```diff
@@ -140,13 +140,14 @@ extern "C" {
 
 /* As defined for TLS1.3 */
-# define TLSEXT_TYPE_key_share                   40
+# define TLSEXT_TYPE_key_share                   51
```
Reorder this to after `TLSEXT_TYPE_signature_algorithms_cert`?

kaduk commented:
Applied review comments and plumbed the TLS 1.3 SignatureScheme values through SSL_CTX_set1_sigalgs_list() to the actual SSL_CTX/SSL objects.
Still needs squashing and some proxy tests to exercise the new extension.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2363,8 +2438,27 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
+            if (s->s3->tmp.peer_cert_sigalgs != NULL) {
+                for (i = 0; i < s->s3->tmp.peer_cert_sigalgslen; i++) {
+                    lu = tls1_lookup_sigalg(s->s3->tmp.peer_cert_sigalgs[i]);
+                    if (lu == NULL || !ssl_has_cert(s, lu->sig_idx))
```
Doesn't this check that the *keys* in the certificate are of the specified type, rather than checking that the *signatures* in the certificate are of the specified type

mattcaswell replied:
Also we don't currently check the signatures for the chain at all at the moment. Should we?

kaduk replied:
We probably should check signatures in the chain, yes.  (Though IIRC there is still a cop-out in the spec that lets you send anything if you can't comply with the peer's request, as a "better-than-nothing" sort of effort.)
And yes, this does check the key in the cert rather than the signature on the cert.  (I don't actually know how to check the latter off the top of my head.)
It seems like there is then a pretty complicated interaction between the signature_algorithms and signature_algorithms_cert processing, in that we have to pick a (leaf) cert that has a key that is compatible with the signature_algorithm that we'll negotiate.  Unless we want to claim that that's an error by the peer and it's okay to fail the connection when we try to perform an operation that can't work.

mattcaswell replied:
> I don't actually know how to check the latter off the top of my head.

`X509_get_signature_info()`?

mattcaswell commented:
Aside: I note that section 4.4.2.2 of draft-23 does not mention "signature_algorithms_cert" at all at the moment. Should it?

mattcaswell commented:
It's not clear to me from the spec whether "signature_algorithms_cert" is mandatory-to-implement. If not, I'm wondering whether it would be better to split this PR into two, so we can uplift to draft-23 now, and add "signature_algorithms_cert" support later.

mattcaswell commented:
What happens if we select a certificate that matches "signature_algorithms_cert", but the keys contained within it don't match "signature_algorithms"?

mattcaswell commented:
> It's not clear to me from the spec whether "signature_algorithms_cert" is mandatory-to-implement.

Ah - its not listed in section 9.2 - which would suggest it is not mandatory to implement.

kaduk commented on ssl/t1_lib.c:
```diff
@@ -1584,19 +1597,22 @@ int tls1_save_u16(PACKET *pkt, uint16_t **pdest, size_t *pdestlen)
 {
     /* Extension ignored for inappropriate versions */
-    if (!SSL_USE_SIGALGS(s))
+    if (!SSL_USE_SIGALGS(s) || (cert && !SSL_IS_TLS13(s)))
```
Good question.  I must have been confused when writing this the first time.

kaduk commented on ssl/t1_lib.c:
```diff
@@ -1722,36 +1740,50 @@ static int sig_cb(const char *elem, int len, void *arg)
+    /* Reject duplicates */
+    for (i = 0; i < sarg->sigalgcnt - 1; i++) {
+        if (sarg->sigalgs[i] == sarg->sigalgs[sarg->sigalgcnt])
             return 0;
```
I guess so; I thought that the 'return 0' would cause the CONF_parse_list() to fail and the whole thing to bail out, but apparently that's not actually the case.
(Also, s_server is happy to receive duplicates, which perhaps it should start enforcing.)

mattcaswell commented:
Ping? What is the status of this?

kaduk commented:
I was reviewing the stateless cookie stuff and didn't have much time for this.
If you are looking for things to do, feel free to pick up the remaining pieces (but please say you're doing so here so that we can avoid duplicating work).

mattcaswell commented:
I'm not looking for things to do!!

kaduk commented:
Added the proxy test for sending signature_algorithms_cert and addressed feedback (hopefully); taking out of WIP.  (Still needs squash+rebase, but holding off in case you want to just look at the new stuff.)

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2302,14 +2399,9 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
-            if (!ssl_has_cert(s, lu->sig_idx)) {
-                if (lu->sig_idx != SSL_PKEY_RSA_PSS_SIGN
-                        || !ssl_has_cert(s, SSL_PKEY_RSA))
-                    continue;
```
I'm not seeing this original logic reflected in `has_usable_cert()`. Shouldn't it be?

kaduk replied:
You are correct that it is not reflected in has_usable_cert().  It is also correct for this logic to not be reflected in has_usable_cert().  This code is in the ```SSL_IS_TLS13()``` block, but it represents the compatibility fallback wherein we let rsaEncryption keys generate PSS signatures.  The whole point of the new codepoints in draft-23 is that we should not do this fallback, since we now have separate codepoints for "PSS signature using rsaEncryption key" and "PSS signature using PSS key".  I left this fallback in for the TLS 1.2 case since I thought there was a way for an old client that only does TLS 1.2 to be trying the PSS signature but not know about the split codepoints.  But maybe that's not actually true so that logic should also be removed...

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2302,14 +2399,9 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
-                if (lu->sig_idx != SSL_PKEY_RSA_PSS_SIGN
-                        || !ssl_has_cert(s, SSL_PKEY_RSA))
-                    continue;
-                sig_idx = SSL_PKEY_RSA;
```
`sig_idx` now never gets set, which means the conditional logic inside the `EVP_PKEY_RSA_PSS` case below never gets hit.

kaduk replied:
good point; I'll remove that, too.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2463,8 +2559,10 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
         sig_idx = lu->sig_idx;
-    s->s3->tmp.cert = &s->cert->pkeys[sig_idx];
-    s->cert->key = s->s3->tmp.cert;
+    if (s->s3->tmp.cert == NULL) {
```
Is this change still correct? I'm not sure why its needed.

kaduk replied:
It is not needed, though it should also not be incorrect (since ```s->s3->tmp.cert``` should always be NULL).

mattcaswell commented on test/recipes/70-test_sslsigalgs.t:
```diff
@@ -314,3 +341,34 @@ sub modify_sigalgs_filter
+            } elsif ($testtype == SIGALGS_CERT_ED448) {
+                $sigs = pack "C4", 0x00, 0x02,
+                              0x08, 0x08;
+            }
```
A comment above each of these explaining what the sigs actually contain in detail would be helpful

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2266,6 +2324,45 @@ static int tls12_get_cert_sigalg_idx(const SSL *s, const SIGALG_LOOKUP *lu)
+             * TODO this does not differentiate between the
+             * rsa_pss_pss_* and rsa_pss_rsae_* schemes since we do not
+             * have a chain here that lets us look at the key OID in the
+             * signing certificate.
```
Right! I understand. Should this TODO actually be TODO(TLS1.3) for consistency with our other TODOs?

kaduk replied:
My thinking was to leave it as just "TODO", since it's unclear that we really need to address it before we consider TLS 1.3 support (and 1.1.1-release) finalized.  (Given that there is a giant loophole in the spec for what certificate chain the server sends back.)

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -2330,10 +2418,7 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
+                pkey = s->cert->pkeys[lu->sig_idx].privatekey;
                 pkey_id = EVP_PKEY_id(pkey);
                 if (pkey_id != EVP_PKEY_RSA_PSS
                     && pkey_id != EVP_PKEY_RSA)
```
Isn't this test irrelevant now?

kaduk replied:
Probably, yes.

mattcaswell commented on test/recipes/70-test_sslsigalgs.t:
```diff
@@ -314,3 +341,39 @@ sub modify_sigalgs_filter
+                             0x04, 0x01,  0x05, 0x01,  0x06, 0x01,  0x02, 0x01;
+            } elsif ($testtype == SIGALGS_CERT_ED448) {
+                $sigs = pack "C4", 0x00, 0x02,
+                             # ed448
```
side note: Might need an update if/when we get ed448 support in

kaduk replied:
True.  I guess I could also make up some non-registered codepoint to send and that would work just as well for the purpose of the test.

kaduk commented:
I'll plan to rebase+squash for the next update.

kaduk commented:
Okay, the rebased+squashed version is up.  (There were some 'make update' conflicts to address.)

kaduk commented:
Thanks for the review!
Looks like we missed an uninitialized read in the duplicate-detection logic in sig_cb that the asan build found, so follow-up coming in a separate pull request.
